data_science/terraform/dlami_asg/userdata.sh.tpl

```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+
+# Create jupyter user
+adduser ${notebook_user} --gecos "" --disabled-password
+
+# Write config file
+mkdir /home/${notebook_user}/.jupyter
+cat << EOF > /home/${notebook_user}/.jupyter/jupyter_notebook_config.py
+
+from s3contents import S3ContentsManager
+
+c = get_config()
+
+# Tell Jupyter to use S3ContentsManager for all storage.
+c.NotebookApp.contents_manager_class = S3ContentsManager
+c.S3ContentsManager.bucket = "${bucket_name}"
+
+c.NotebookApp.notebook_dir = u"/home/${notebook_user}/"
+c.NotebookApp.ip = "*"
+c.NotebookApp.port = ${notebook_port}
+c.NotebookApp.open_browser = False
+c.NotebookApp.password = u'${hashed_password}'
+
+EOF
+
+pip install pillow \
+            seaborn \
+            scikit-learn \
+            tqdm
+
+pip install jupyter \
+            boto3 \
+            requests \
+            s3fs==0.1.2 \
+            mock \
+            nose
```

Comment:
If we just install the latest version of every package, at some point one of these packages will break and we'll have a sad Harrison. It would be better to install a pinned set of packages that changes _only at our convenience_.

This is why we use pip-tools everywhere else: https://github.com/jazzband/pip-tools

It would be nice to do the same here (although how we get a requirements.txt into this script I'm not sure).

Reply:
Yup, this approach using userdata to run a bootstrap script is tricky as soon as you want to do anything complex, so we should probably just pin them here. I want to avoid building our own AMI as long as I can because the dev loop on that is tedious (as you have to wait for the ami to build in AWS for every change).